### DISCUSSION TITLE
Loaders are not re-called if an action returns a redirect

### DISCUSSION THREAD

**Discussion by charklewis** [remix-run/remix#1944] [Q&A] [answered]
I have noticed unexpected behaviour when I return `redirect` from an action. I have situations where a success in an action results in a redirect. However, when I do this, the loaders don't get re-called like they do if I return the action with a `2XX` status code (eg if I use `json`). This results in stale data in the UI. 

There are many situations in my app(s) where the user perform tasks of the following nature:

1. Be on their profile page
2. Click on a button to create an item
3. This will call an action and when successful redirect them to that item's page
4. The user will then click a button in the UI to go back to their profile to see a list of all of their items, however the new item is missing (due to the loaders never being re-called). 

A simple fix I have found is to return `json` in the loader and use an `useEffect` to detect the success response from the action and then redirect the user. This works perfectly. However, it feels like something I should be able to do from an action. 

A check I have done is to return `2XX` in the `redirect` function. I found this re-calls the loaders (so the UI updates correctly). It just doesn't redirect to the new page (as expected). So its seems clear that it is an issue with the status code. 

I have also tested with the below to ensure they aren't causing the issue:

- Removed all prefetching
- Disabled all caches in my browser
- Added `no-cache` or `max-age=0` to all responses

I might have missed something in the docs, I can't seem to see anywhere that describes how the re-calling of loaders works with different status codes.

**Comment by rossipedia** [comment]
This feels more like an issue than a discussion?

**Reply by charklewis** [reply]
I wasn't sure if it was expected behaviour or not. If it was expected behaviour then suggestions on how to best handle.

**Comment by sergiodxa** [comment]
> The user will then click a button to go back

Is this back button the browser back button or just a link pointing to the profile?

**Reply by charklewis** [reply]
I haven't tried clicking back in the browser. My testing has just been using a button in the UI. There are normally a few places in the UI that a user can click to navigate to their profile, they all have the same behaviour (ie from a drop down menu, breadcrumbs, "go back to profile" button etc).

**Reply by charklewis** [reply]
Do you think this could be a bug @sergiodxa ?

**Comment by charklewis** [comment, marked as answer]
It appears there is an issue marked as a bug with similar behaviour - https://github.com/remix-run/remix/issues/1616
